src/java/org/apache/cassandra/cql3/Terms.java

```diff
@@ -47,225 +47,510 @@ public Object get(int index)
-    {
-        return new Terms()
+                @Override
+                public List<ByteBuffer> get()
                 {
-                    @Override
-                    public void addFunctionsTo(List<Function> functions)
+                    int size = terminals.size();
+                    List<ByteBuffer> buffers = new ArrayList<>(size);
+                    for (int i = 0; i < size; i++)
                     {
-                        term.addFunctionsTo(functions);
+                        Terminal terminal = terminals.get(i);
+                        buffers.add(terminal == null ? null : terminal.get());
                     }
+                    return buffers;
+                }
 
-                    @Override
-                    public void collectMarkerSpecification(VariableSpecifications boundNames)
+                @Override
+                public List<List<ByteBuffer>> getElements()
+                {
+                    int size = terminals.size();
+                    List<List<ByteBuffer>> buffers = new ArrayList<>(size);
+                    for (int i = 0; i < size; i++)
                     {
-                        term.collectMarkerSpecification(boundNames);
+                        Terminal terminal = terminals.get(i);
+                        buffers.add(terminal == null ? null : terminal.getElements());
                     }
+                    return buffers;
+                }
 
-                    @Override
-                    public List<ByteBuffer> bindAndGet(QueryOptions options)
-                    {
-                        return Collections.singletonList(term.bindAndGet(options));
-                    }
+                @Override
+                public void addFunctionsTo(List<Function> functions)
+                {
+                    addFunctions(terminals, functions);
+                }
 
-                    @Override
-                    public List<Terminal> bind(QueryOptions options)
-                    {
-                        return Collections.singletonList(term.bind(options));
-                    }
-                };
+                @Override
+                public boolean containsSingleTerm()
+                {
+                    return terminals.size() == 1;
+                }
+
+                @Override
+                public Terminal asSingleTerm()
+                {
+                    if (!containsSingleTerm())
+                        throw new UnsupportedOperationException("This terms contains cannot be converted in a single term");
+                    return terminals.get(0);
+                }
+            };
+        }
     }
 
-    /**
-     * Creates a {@code Terms} containing a set of {@code Term}.
-     *
-     * @param term the {@code Term}
-     * @return a {@code Terms} containing a set of {@code Term}.
-     */
-    public static Terms of(final List<Term> terms)
+    abstract class NonTerminals implements Terms
```

Comment:
It might be useful to clarify in the class JavaDoc whether it's expected that all the terms in this class are expected to be non-terminal, or if it might contain at least one non-terminal.

Reply:
I understand it's the latter, and there could be terminal terms in a `NonTerminals`. So maybe the class could be named `SomeNonTerminals` or something like that? A clear JavaDoc is probably enough if we want to keep the current name.

Reply:
I updated the doc.